plugins/network-elements/nicira-nvp/src/main/java/com/cloud/network/resource/wrapper/NiciraNvpConfigureSharedNetworkUuidCommandWrapper.java

```diff
@@ -0,0 +1,110 @@
+import com.cloud.resource.ResourceWrapper;
+
+@ResourceWrapper(handles =  ConfigureSharedNetworkUuidCommand.class)
+public final class NiciraNvpConfigureSharedNetworkUuidCommandWrapper extends CommandWrapper<ConfigureSharedNetworkUuidCommand, Answer, NiciraNvpResource>{
+
+    private static final Logger s_logger = Logger.getLogger(NiciraNvpConfigureSharedNetworkUuidCommandWrapper.class);
+
+    @Override
+    public Answer execute(ConfigureSharedNetworkUuidCommand command, NiciraNvpResource niciraNvpResource) {
+        final String logicalRouterUuid = command.getLogicalRouterUuid();
+        final String logicalSwitchUuid = command.getLogicalSwitchUuid();
+        final String portIpAddress = command.getPortIpAddress();
+        final List<NiciraNvpTag> tags = new ArrayList<NiciraNvpTag>();
+        tags.add(new NiciraNvpTag("cs_account", command.getOwnerName()));
+        final long networkId = command.getNetworkId();
+
+        final NiciraNvpApi niciraNvpApi = niciraNvpResource.getNiciraNvpApi();
+
+        s_logger.debug("Attaching Logical Switch " + logicalSwitchUuid + " on Logical Router " + logicalRouterUuid + " for Shared Network" + networkId);
+
+        //Stored for rollback
+        LogicalRouterPort lrpi = null;
+        LogicalSwitchPort lsp = null;
+        try {
+            // Create the inside port for the router
+            lrpi = new LogicalRouterPort();
+            lrpi.setAdminStatusEnabled(true);
+            lrpi.setDisplayName(niciraNvpResource.truncate(networkId + "-shared-att-port", NAME_MAX_LEN));
+            lrpi.setTags(tags);
+            final List<String> ipAddresses = new ArrayList<String>();
+            ipAddresses.add(portIpAddress);
+            lrpi.setIpAddresses(ipAddresses);
+            lrpi = niciraNvpApi.createLogicalRouterPort(logicalRouterUuid, lrpi);
+
+            try {
+                // Create the inside port on the lswitch
+                lsp = new LogicalSwitchPort(niciraNvpResource.truncate(networkId + "-shared-att-port", NAME_MAX_LEN), tags, true);
+                lsp = niciraNvpApi.createLogicalSwitchPort(logicalSwitchUuid, lsp);
+
+                // Attach the inside router port to the lswitch port with a PatchAttachment
+                niciraNvpApi.updateLogicalRouterPortAttachment(logicalRouterUuid, lrpi.getUuid(), new PatchAttachment(lsp.getUuid()));
+
+                // Attach the inside lswitch port to the router with a PatchAttachment
+                niciraNvpApi.updateLogicalSwitchPortAttachment(logicalSwitchUuid, lsp.getUuid(), new PatchAttachment(lrpi.getUuid()));
+            }
+            catch (final NiciraNvpApiException e){
+                try {
+                    niciraNvpApi.deleteLogicalRouterPort(logicalRouterUuid, lrpi.getUuid());
+                    if (lsp != null){
+                        niciraNvpApi.deleteLogicalSwitchPort(logicalSwitchUuid, lsp.getUuid());
+                    }
+                } catch (NiciraNvpApiException ex) {
```

Comment:
This is indeed better. However, now you are logging it as an info message, and then making a new exception.
If you log something, and then throw an exception, I would say the log level should be warn because you are delegating the handling of the exception to the callers.
In addition, when you make a new exception out of an existing exception you should include the original exception as the cause of the new exception.
In this case you are using two exception as the cause `e and ex`. 

Let us think of a way of doing this better:
* The first exception `ex` is being handled here. So I would say, first thing is to log it as an error (`s_logger.error("failed to do something", ex)`). This way we explain what failed and present the cause in the logs.
* Then we handle that exception (`ex`) and we might get a second one (`e`). 
* This second one we don't want to handle here, and we will bubble it up wrapped in a new exception. So then we log `e` as a warning (`s_logger.warn("failed to do something else, namely handing the previous excpetion", e)`), and then we wrap it in  a new exception and throw it (`throw new NiciraNvpApiException("probably the same message we set in the warn log", e`).

What do you think? Does this make sense to you?